packages/scoped-elements/README.md

```diff
@@ -0,0 +1,276 @@
+
+## How it works
+
+`createScopedHtml` is going to define the custom element classes and creates a
+unique scoped tag for each of them if needed. In case those classes were
+previously defined then it's going to return the previous defined tag.
+
+Then, the `html` function provided by the `createScopedHtml` method is going to
+transform the template literal into another one replacing the tags used by the
+user by the ones defined by the custom elements. Finally, the transformed
+template literal is going to be processed by `lit-html`.
+
+`<my-button>${this.text}</my-button>` --> `<my-button-3>${this.text}</my-button-3>`
+
+## How to use it
+
+1. Import the `createScopedHtml` from `@open-wc/scoped-elements`.
+
+   ```js
+   import { createScopedHtml } from '@open-wc/scoped-elements';
+   ```
+
+2. Import the classes of the components you want to use.
+
+   ```js
+   import MyButton from './MyButton.js';
+   import MyPanel from './MyPanel.js';
+   ```
+
+3. Create the `html` function defining the tags you want to use for the
+   components.
+
+   ```js
+   const html = createScopedHtml({
+     'my-button': MyButton,
+     'my-panel': MyPanel,
+   });
+   ```
+
+4. Use your components in your html.
+
+   ```js
+   render() {
+     return html`
+       <my-panel class="panel">
+         <my-button>${this.text}</my-button>
+       </my-panel}>
+     `;
+   }
+   ```
+
+### Complete example
+
+```js
+import { css, LitElement } from 'lit-element';
+import { createScopedHtml } from '@open-wc/scoped-elements';
+import MyButton from './MyButton.js';
+import MyPanel from './MyPanel.js';
+
+const html = createScopedHtml({
+  'my-button': MyButton, // <-- bind the element with a tag in your html
+  'my-panel': MyPanel,
+});
+
+export default class MyElement extends LitElement {
+  static get styles() {
+    return css`
+      .panel {
+        padding: 10px;
+        background-color: grey;
+      }
+    `;
+  }
+
+  static get properties() {
+    return {
+      text: String,
+    };
+  }
+
+  render() {
+    return html`
+      <my-panel class="panel">
+        <my-button>${this.text}</my-button>
+      </my-panel}>
+    `;
+  }
+}
+```
+
+## Limitations
+
+1. Components imported via npm **MUST NOT** be self registering components
+   If a shared component (on npm) does not offer an export to the class alone, without the registration side effect, then this component can not be used. e.g. every component
+   that calls `customElement.define`
+
+   ```js
+   export class MyEl { ... }
+   customElement.define('my-el', MyEl);
+   ```
+
+   or uses the typescript decorator `customElement`
+
+   ```ts
+   @customElement('my-el')
+   export class MyEl { ... }
+   ```
+
+   only side effects free class exports may be used
+
+   ```js
+   export class MyEl { ... }
+   ```
+
+   Note: not share or components withing the main app shell are exempted from this rule.
```

Comment:
What does this mean? "Not share or components" ? Can we rewrite this?

Reply:
It should mean nothing as it can be total ignored he was trying to point out that if there is already a component defined with a tagname then this will try to register it also which has no negativ impact at all.

Reply:
Then it should be removed

Reply:
How about

Note: not shared components or those withing the main app shell are exempted from this rule.

Reply:
What exactly do you mean with shared components? Isnt that the point of this library?

Reply:
mmm probably it's a bit confusing. Maybe "Those components that are not going to be shared in a library are exempted from this rule."

Reply:
Ok maybe i am missing something but as soon as this is merged i could add some extras from my code base i  for example check if the element that i get is the same class that i already wanted to register to dedub registration and this 

registration helper should not care where dependencys come from global dependencys are most time self versioning via namespaces look for example into lit-html it places a lit version var on the global scope

in this registration component you can completly ignore what your users are doing this has only one goal 

RREGISTER A NEW Component under a Uniq NAME or Return the Already Registered thats it all other dependes to the developers choice if he chooses to use global dependencys he is free to do so this will never replace scoped registry and scoped registry will also never handle global modules 

as scoped registry is not Isolate Components 

a pattern for isolated Components with a individual global scope can be also supplyed via IFRAME it adds a extra window object to the script that it can use and then we could deepCopy the Elements i have even a algo that does copy the event listners and all that which clone node and import node would ignore.

Reply:
Ah Maybe the section should point out that this helper is best used with customElements that are coming from a 3th party ???? to make sure they don't  use a already registered tagName?

We Could also add some other interristing part a Loader !

This could even load a customElement and shim customElement calls via setting a scoped customElements Proxy object it don't needs to be a real JS Proxy but i choose this name as it does good explain what it should do it should simply reference this helper as customElement.define()

most self registering modules will have the line customElements.define()
if we wrapp that into import() call where the function that does the call has a local customElements Proxy it will get used!.

Other method is to use for example my esm-loader and simply remove calls to customElements.define.

Reply:
Ok, I think I'm misunderstanding something and also I apologize because my English level doesn't help to explain myself.

This note is referring to the first point.
>1. Components imported via npm **MUST NOT** be self registering components. If a shared component (installed from npm) does not offer an export to the class alone, without the registration side effect, then this component can not be used. E.g. every component that calls `customElement.define`

What it means is that if you are building a component that is susceptible to be used by third parties (which basically means that is going to be imported by other packages), you must **export the component class without registering the component in the registry**, so the third party would be able to use this feature registering the component safely.

If you are creating your components in the same package as the application shell then, probably, you are not going to share or distribute them, so you could register them in the same way you always do.

>Ok maybe i am missing something but as soon as this is merged i could add some extras from my code base i for example check if the element that i get is the same class that i already wanted to register to dedub registration and this

@frank-dspeed sorry but I'm not sure if I'm understanding you well. This feature is checking if the class or the tag are already registered. Each custom element class is going to be registered only once, and its tag is going to be reused by all their references.

Reply:
@manolakis that is not correct for example when some one does not bundle and is using components like
<script src="" type="module">
<script src="" type="module">
<script src="" type="module">
your map will not get reused as its scoped to the module context not the global context all 2 modules will register a indipendent version. of the Same Class 

so your code needs customElements.get()  === Klass 

to verify if it is already the same class that gets returned.

Reply:
also it would be a good idea to drop the scope registration in favor of a uuid tagname and a global weak map the weakmap should get any uniqe name could be even a cryptic one to make sure it will never exist by any other module then this can get reused over all instances to make sure that registration happens only once for a customElement version that is all it is made for 

get uniqe tagname vor a customElement version based on the global context the customElement version can be validated via 2 ways first way ist customElement.get() === Klass and the secund more effective is class.toString() === customElements.get().toString() 
and then clean up that string so that white space and that don't matters as long as the code is the same content.

Reply:
Its importend maybe to point out that ESM has a global module registry that means when you import a from 'a' in one script and import it in a other you get the same module.

Reply:
Importing the same module twice will return the same module, that's one of the main points of modules.

Reply:
Alright, I completely lost the plot here 😄 Can we try and keep the discussions somewhat focussed?